### PR TITLE
Log

### PR DESCRIPTION
**Pull request by ebuchman**
Combines https://github.com/tendermint/tmlibs/pull/1 and https://github.com/tendermint/tmlibs/pull/5. See the discussion in https://github.com/tendermint/tmlibs/pull/5

### CODE REVIEW

File: common/service.go [hunk 1]
```diff
@@ -19,6 +19,8 @@ type Service interface {
 	IsRunning() bool
 
 	String() string
+
+	SetLogger(log.Logger)
```
Comment (ebuchman) [inline common/service.go]: Can u remind me why the constructors are taking loggers if we have this method ?

File: log/filter.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+	for _, option := range options {
+		option(l)
+	}
+	return l
+}
+
+type filter struct {
+	next          Logger
+	allowed       level
+	errNotAllowed error
```
Comment (ebuchman) [inline log/filter.go]: what makes this not nil?

File: common/service.go (hunk 1 above)
Comment (melekes) [inline common/service.go]: they should not. but in single cases where we log something in a constructor, there is no other way

File: log/filter.go (hunk 2 above)
Comment (melekes) [inline log/filter.go]: option below ErrNotAllowed

File: log/filter.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+func AllowAll() Option {
+	return AllowDebug()
+}
+
+// AllowDebug allows error, warn, info and debug level log events to pass.
+func AllowDebug() Option {
+	return allowed(levelError | levelInfo | levelDebug)
+}
+
+// AllowInfo allows error, warn and info level log events to pass.
```
Comment (ebuchman) [inline log/filter.go]: warn? (and above)

File: log/filter.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+func allowed(allowed level) Option {
+	return func(l *filter) { l.allowed = allowed }
+}
+
+// ErrNotAllowed sets the error to return from Log when it squelches a log
+// event disallowed by the configured Allow[Level] option. By default,
+// ErrNotAllowed is nil; in this case the log event is squelched with no
+// error.
+func ErrNotAllowed(err error) Option {
```
Comment (ebuchman) [inline log/filter.go]: I see. so callers need to call this if they want the err to not be nil

Comment (melekes) [inline log/filter.go]: `log.NewFilter(logger, log.ErrNotAllowed(errors.New("hey")))`

https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis

Comment (ebuchman) [inline log/filter.go]: love it

**Pull request merged**